### ISSUE TITLE
DefineMap/DefineSet/DefineIn annotation

### ISSUE THREAD

**Issue by mdhausman** [jdbi/jdbi#512]
I raised this in #490 and it was suggested that I create a new issue for it, so here it is:

> I recently had a use-case in which we needed to `@Define` a list of variable length.
> 
> Essentially, the statement was
> 
> ``` sql
> INSERT INTO <tableName>
>   (<fieldNames>)
> VALUES
>   (<fieldValues>)
> ```
> 
> But with no way to `@Define` a list of fieldNames, I had to write up a custom `@DefineIn` annotation to accept a `Set<String>` argument. It worked well, but it seems like it would be a good feature for JDBI3.

**Comment by mdhausman** [comment]
I'd be happy to write up a little PR with my annotation if it's desired.

**Comment by stevenschlansker** [comment]
Yes, if you could PR your annotation plus tests, that would be fantastic!

**Comment by qualidafial** [comment]
Before you start work, could you briefly describe the solution you came up with, and give an example of its usage?

I wonder whether `Set` is the right abstraction--sets are not guaranteed to be ordered, and order is significant in queries like your example `insert` statement. Maybe once I see its usage it won't be a problem--but I have to see it to understand what your intent is.

As an aside, this issue made me think of some possible convenience methods we could add to `Handle`:

    void insert(String tableName, Map<String, Object> values);
    List<Map<String,Object>> select(String tableName, String... columnNames);
    void update(String tableName, Map<String, Object> identifiers, Map<String, Object> values);
    void delete(String tableName, Map<String, Object> identifiers);

Not sure how useful these would be in practice--just an idea.

**Comment by stevenschlansker** [comment]
Good point.  `List` might be more appropriate.  If there are some use cases where order doesn't matter, you might consider just using `Collection` or even `Iterable` and leaving it up to the end user whether order matters -- but understanding use cases is important.

**Comment by mdhausman** [comment]
To be honest, I actually have it taking a Collection in my code because it's just a little custom annotation for the project. I'd said Set here because I thought that in hindsight a set might make more sense; you can't have two columns of the same name in any SQL I've ever heard of, and having multiple identical members would cause SQL exceptions.

But you can only user-proof something so much.

**Comment by mdhausman** [comment]
The solution is basically just iterating through the arg and writing the members into a comma separated list. It's nothing terribly fancy, but useful to have abstracted into an annotation. 

It's been working well on our end, where we've used it to replace several SQL insert statements with a single yaml-driven generic insert statement. So instead of having to touch several different files to start streaming data into a new table, we just modify a single yaml.

**Comment by mdhausman** [comment]
The business end of mine looks something like this:

```java
@Override
public SqlStatementCustomizer createForParameter(Annotation annotation, Class sqlObjectType, Method method, final Object arg) {
  DefineIn d = (DefineIn) annotation;
  final String key = d.value();
  return new SqlStatementCustomizer() {
    @Override
    public void apply(SQLStatement q) {
      if (arg instanceof Collection) {
        Collection<?> argsList = (Collection<?>) arg;
        StringBuilder sb = new StringBuilder();
        boolean firstItem = true;
        for (Object o : argsList) {
          if (firstItem) {
            firstItem = false;
          } else {
            sb.append(',');
          }
          sb.append(o.toString());
        }
        q.define(key, sb.toString());
      } else {
        throw new UnsupportedOperationException("Only supported on Collection arguments");
      }
    }
  };
}
```

Though for an open-source release I'll want to clean that up a bit (not to mention write some more tests--I've been looking over JDBI's existing tests for guidelines). Insertion order should be important, as that would determine the order in which one would bind values. This would suggest a List rather than a Collection. But no SQL is going to allow two columns of the same name in the same table, so there might be value in requiring unique members. This would suggest a LinkedHashSet. But that's a bit more limiting, and might be less useful; if users are getting SQL exceptions from putting duplicate entries in their List of columns, that's no fault of the library.

In use, it looks something like this:

```java
@SqlUpdate("INSERT IGNORE INTO <schema>.<tableName> (<columns>) VALUES (<values>)")
int dynamicSqlInsert(
  @Define("schema") String schema,
  @Define("tableName") String tableName,
  @DefineIn("columns") List<String> columns,
  @BindIn("values") List<Object> values
);
```

The columns and values are dynamically generated, fieldNames coming from a YAML list and fieldValues being streamed out of Kafka.

**Comment by qualidafial** [comment]
What about a `Map<String,Object>`? e.g.

    @SqlUpdate("INSERT IGNORE INTO <schema>.<tableName> (<columns>) VALUES (<values>)")
    int dynamicSqlInsert(
      @Define("schema") String schema,
      @Define("tableName") String tableName,
      @DefineKeysIn("columns") @BindValuesIn("values") Map<String, Object> values
    );

This way iteration order doesn't matter, as long as it is consistent.

**Comment by qualidafial** [comment]
With that said, I'm not opposed to `@DefineIn` using a list of columns, the way you've described it.

**Comment by mdhausman** [comment]
Though I think it might be better for the Map annotation to function as a single annotation. Something like `@BindValueMap("columns", "values") Map<String, Object> valueMap`

**Comment by qualidafial** [comment]
It would have to be something like `@BindValueMap(keys="columns", values="values")`, but yes I considered something like that too.

I think we're basically on the same page. If you want to run the ball on this and put together a PR, that would be great. Otherwise I'll get to this as soon as I can.

**Comment by mdhausman** [comment]
Hah, I'm pretty sure I could only get company time justification for putting out `@DefineIn` (since I already wrote it for the company). I could look into setting up a separate one for a `@BindValueMap` on my own time after that, though. Would that work?

**Comment by qualidafial** [comment]
I'll accept all the free help I can get ;)

Let's plan on `@DefineIn` first, then we'll follow up later for the `@DefineMap` or `@BindValueMap` or whatever we decide to call it.

Thanks Maura

**Comment by qualidafial** [comment]
Fixed in #533
